src/python_testing/matter_testing_support.py

```diff
@@ -436,55 +451,52 @@ def as_string(self, mapper: ClusterMapper):
         return desc
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Attribute:{id_str(self.attribute_id)}')
+        return (f'{super().__str__()}'
+                f'\n      Attribute:{id_str(self.attribute_id)}')
 
 
 @dataclass
-class EventPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class EventPathLocation(ClusterPathLocation):
     event_id: int
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Event:    {id_str(self.event_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Event:    {id_str(self.event_id)}')
 
 
 @dataclass
-class CommandPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class CommandPathLocation(ClusterPathLocation):
     command_id: int
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Command:  {id_str(self.command_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Command:  {id_str(self.command_id)}')
 
 
 @dataclass
-class ClusterPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class FeaturePathLocation(ClusterPathLocation):
+    feature_code: str
 
     def __str__(self):
-        return (f'\n       Endpoint: {self.endpoint_id},'
-                f'\n       Cluster:  {cluster_id_str(self.cluster_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Feature:  {self.feature_code}')
 
 
 @dataclass
-class FeaturePathLocation:
-    endpoint_id: int
-    cluster_id: int
-    feature_code: str
+class DeviceTypePathLocation(ProblemLocation):
+    device_type_id: int
+    cluster_id: Optional[int]
+
+    def __init__(self, device_type_id: int, cluster_id: Optional[int] = None):
+        self.device_type_id = device_type_id
+        self.cluster_id = cluster_id
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Feature:  {self.feature_code}')
+        msg = f'\n       DeviceType: {self.device_type_id}'
```

Comment:
existing code, however I would note that str usuaully does not start with \n. It may be better to have some methods like `Format` if we want custom formatting rather than trying to use str or repr.

Reply:
Going to leave this for now - if we want to muck about with string formatting, that's fine, but in a separate PR. There's a lot of existing code here.

Reply:
@andy31415 - LMK if that's ok. I think before we push up as a package we'll just need a bit of general cleanup here and should handle that in one editorial PR.